Source/Core/Core/HW/DVDThread.cpp

```diff
@@ -0,0 +1,188 @@
+static Common::Event s_DVD_thread_start_working;
+static Common::Event s_DVD_thread_not_working;
+static Common::Flag s_DVD_thread_exiting(false);
+
+static std::vector<u8> s_DVD_buffer;
+static u64 s_time_read_started;
+static bool s_DVD_success;
+
+static u64 s_DVD_offset;
+static u32 s_output_address;
+static u32 s_length;
+static bool s_decrypt;
+
+// Used to notify emulated software after executing command.
+// Pointers don't work with savestates, so CoreTiming events are used instead
+static int s_callback_event_type;
+
+// The following time variables are only used for logging
+static u64 s_realtime_started_us;
+static u64 s_realtime_done_us;
+
+void Start()
+{
+	s_finish_read = CoreTiming::RegisterEvent("FinishReadDVDThread", FinishRead);
+	_assert_(!s_DVD_thread.joinable());
+	s_DVD_thread = std::thread(DVDThread);
+}
+
+void Stop()
+{
+	_assert_(s_DVD_thread.joinable());
+
+	// The DVD thread will return if s_DVD_thread_exiting
+	// is set when it starts working
+	s_DVD_thread_exiting.Set();
+	s_DVD_thread_start_working.Set();
+
+	s_DVD_thread.join();
+
+	s_DVD_thread_exiting.Clear();
+}
+
+void DoState(PointerWrap &p)
+{
+	WaitUntilIdle();
+
+	// TODO: Savestates can be smaller if s_DVD_buffer is not saved
+	p.Do(s_DVD_buffer);
+	p.Do(s_time_read_started);
+	p.Do(s_DVD_success);
+
+	p.Do(s_DVD_offset);
+	p.Do(s_output_address);
+	p.Do(s_length);
+	p.Do(s_decrypt);
+	p.Do(s_callback_event_type);
+
+	// s_realtime_started_us and s_realtime_done_us aren't savestated
+	// because they rely on the current system's time.
+	// This means that loading a savestate might cause
+	// incorrect times to be logged once.
+}
+
+void WaitUntilIdle()
+{
+	_assert_(Core::IsCPUThread());
+
+	// Wait until DVD thread isn't working
+	s_DVD_thread_not_working.Wait();
+
+	// Set the event again so that we still know that the DVD thread isn't working
+	s_DVD_thread_not_working.Set();
+}
+
+void StartRead(u64 DVD_offset, u32 output_address, u32 length, bool decrypt,
+               int callback_event_type, int ticks_until_completion)
+{
+	_assert_(Core::IsCPUThread());
+
+	s_DVD_thread_not_working.Wait();
+
+	s_DVD_offset = DVD_offset;
+	s_output_address = output_address;
+	s_length = length;
+	s_decrypt = decrypt;
+	s_callback_event_type = callback_event_type;
+
+	s_time_read_started = CoreTiming::GetTicks();
+	s_realtime_started_us = Common::Timer::GetTimeUs();
+
+	s_DVD_thread_start_working.Set();
+
+	CoreTiming::ScheduleEvent(ticks_until_completion, s_finish_read);
+}
+
+static void FinishRead(u64 userdata, int cyclesLate)
+{
+	WaitUntilIdle();
+
+	DEBUG_LOG(DVDINTERFACE, "Disc has been read. Real time: %llu us. Real time including delay: %llu us. Emulated time including delay: %llu us.",
+	          s_realtime_done_us - s_realtime_started_us,
+	          Common::Timer::GetTimeUs() - s_realtime_started_us,
+	          (CoreTiming::GetTicks() - s_time_read_started) / (SystemTimers::GetTicksPerSecond() / 1000 / 1000));
+
+	while (!s_DVD_success)
```

Comment:
I don't want emulation to continue running when the user is being asked the question. Doing this on the DVD thread isn't especially bad right now, because the CPU thread will stop and wait for the DVD thread after a short time no matter what, but it will matter in the future with my planned option to only call FinishRead when the DVD thread is done.